services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/resource_tracking/core.py

```diff
@@ -0,0 +1,129 @@
+from servicelib.logging_utils import log_context
+
+from ...core.docker_utils import (
+    are_all_containers_in_expected_states,
+    get_container_states,
+)
+from ...core.rabbitmq import post_resource_tracking_message
+from ...core.settings import ApplicationSettings
+from ...models.shared_store import SharedStore
+from ._models import ResourceTrackingState
+from .settings import ResourceTrackingSettings
+
+_STOP_WORKER_TIMEOUT_S: Final[NonNegativeFloat] = 1.0
+
+_logger = logging.getLogger(__name__)
+
+
+def _get_settings(app: FastAPI) -> ApplicationSettings:
+    settings: ApplicationSettings = app.state.settings
+    return settings
+
+
+async def _start_heart_beat_task(app: FastAPI) -> None:
+    settings: ResourceTrackingSettings = app.state.settings.RESOURCE_TRACKING
+    resource_tracking: ResourceTrackingState = app.state.resource_tracking
+
+    if resource_tracking.heart_beat_task is not None:
+        msg = f"Unexpected task={resource_tracking.heart_beat_task} already running!"
+        raise RuntimeError(msg)
+
+    with log_context(_logger, logging.DEBUG, "starting heart beat task"):
+        resource_tracking.heart_beat_task = start_periodic_task(
+            _heart_beat_task,
+            app=app,
+            interval=settings.RESOURCE_TRACKING_HEARTBEAT_INTERVAL,
+            task_name="resource_tracking_heart_beat",
+        )
+
+
+async def stop_heart_beat_task(app: FastAPI) -> None:
+    # NOTE: this is only used by the teardown
+    await __stop_heart_beat_task(app)
+
+
+async def __stop_heart_beat_task(app: FastAPI) -> None:
+    resource_tracking: ResourceTrackingState = app.state.resource_tracking
+    if resource_tracking.heart_beat_task:
+        await stop_periodic_task(
+            resource_tracking.heart_beat_task, timeout=_STOP_WORKER_TIMEOUT_S
+        )
+
+
+async def _heart_beat_task(app: FastAPI):
+    settings: ApplicationSettings = _get_settings(app)
+    shared_store: SharedStore = app.state.shared_store
+
+    container_states: dict[str, ContainerState | None] = await get_container_states(
+        shared_store.container_names
+    )
+
+    if are_all_containers_in_expected_states(container_states.values()):
+        message = RabbitResourceTrackingHeartbeatMessage(
+            service_run_id=settings.DY_SIDECAR_RUN_ID
+        )
+        await post_resource_tracking_message(app, message)
+    else:
+        _logger.info(
+            "heart beat message skipped: container_states=%s", container_states
+        )
+
+
+async def send_service_stopped(
+    app: FastAPI, simcore_platform_status: SimcorePlatformStatus
+) -> None:
+    # NOTE: calling `stop_heart_beat_task` in place of `__stop_heart_beat_task` does not work.
+    # Somehow the function does not get called.
+    # After spending a lot of time on figuring this out, the current is the best I can offer.
+    # If you want to refactor this talk with ANE first before sinking more time in it.
+    await __stop_heart_beat_task(app)
```

Comment:
Not sure I get the why now here. so you say that calling ```stop_heart_beat_task``` and ```__stop_heart_beat_task``` works is that still true or some old comment?

Reply:
I am a bit puzzled with this comment. I wonder whether the problem you detected is caused by using a private function instead of a protected one? Why you add double `_`

Reply:
ok function was mocked, refactored